src/main/java/picard/sam/BamErrorMetric/BaseErrorAggregation.java

```diff
@@ -0,0 +1,88 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package picard.sam.BamErrorMetric;
+
+import htsjdk.samtools.reference.SamLocusAndReferenceIterator.SAMLocusAndReference;
+import htsjdk.samtools.util.CollectionUtil;
+import htsjdk.samtools.util.SamLocusIterator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.TreeSet;
+import java.util.function.Supplier;
+
+/**
+ * An interface and implementations for classes that apply a {@link picard.sam.BamErrorMetric.ReadBaseStratification.RecordAndOffsetStratifier RecordAndOffsetStratifier}
+ * to put bases into various "bins" and then compute an {@link ErrorMetrics} on these bases using a {@link BaseErrorCalculation}.
+ *
+ * @author Yossi Farjoun
+ */
+
+public class BaseErrorAggregation<CALCULATOR extends BaseErrorCalculation.BaseCalculator> {
+    private final Supplier<CALCULATOR> simpleAggregatorGenerator;
+    private final ReadBaseStratification.RecordAndOffsetStratifier stratifier;
+
+    // Updated as a defaulting map.
+    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
+    private final CollectionUtil.DefaultingMap<Object, CALCULATOR> strataAggregatorMap = new CollectionUtil.DefaultingMap<>
+            (new CollectionUtil.DefaultingMap.Factory<CALCULATOR, Object>() {
+                @Override
+                public CALCULATOR make(final Object ignored) {
+                    return simpleAggregatorGenerator.get();
+                }
+            }, true);
+
+    public BaseErrorAggregation(final Supplier<CALCULATOR> simpleAggregatorGenerator,
+                                final ReadBaseStratification.RecordAndOffsetStratifier stratifier) {
+        this.stratifier = stratifier;
+        this.simpleAggregatorGenerator = simpleAggregatorGenerator;
+    }
+
+    public void addBase(final SamLocusIterator.RecordAndOffset recordAndOffset, final SAMLocusAndReference locusInfo) {
+        final Object stratus = stratifier.stratify(recordAndOffset, locusInfo);
+        // this assumes we do not want to aggregate null.
+        if (stratus != null) strataAggregatorMap.get(stratus).addBase(recordAndOffset, locusInfo);
+
+    }
+
+    public String getSuffix() {
+        return simpleAggregatorGenerator.get().getSuffix() + "_by_" + stratifier.getSuffix();
+    }
+
+    public ErrorMetrics.ErrorMetricBase[] getMetrics() {
```

Comment:
I'm not opposed to nested classes but in cases where a class is heavily used outside its parent, it makes for awkward code like this. Could this be moved to a top level type?

Alternatively, you could rename it to `Base` so code that refers to it would now read `ErrorMetrics.Base` which isn't as awkward.